libvast/src/table_slice.cpp

```diff
@@ -399,101 +393,204 @@ std::span<const std::byte> as_bytes(const table_slice& slice) noexcept {
 
 // -- operations ---------------------------------------------------------------
 
-void select(std::vector<table_slice>& result, const table_slice& slice,
-            const ids& selection) {
-  const auto offset = slice.offset() == invalid_id ? 0 : slice.offset();
+table_slice join(std::vector<table_slice> slices) {
+  slices.erase(std::remove_if(slices.begin(), slices.end(),
+                              [](const auto& slice) {
+                                return slice.encoding()
+                                       == table_slice_encoding::none;
+                              }),
+               slices.end());
+  if (slices.empty())
+    return {};
+  if (slices.size() == 1)
+    return std::move(slices[0]);
+  auto schema = slices[0].layout();
+  VAST_ASSERT(std::all_of(slices.begin(), slices.end(),
+                          [&](const auto& slice) {
+                            return slice.layout() == schema;
+                          }),
+              "join requires slices to be homogeneous");
```

Comment:
Should this assertion come earlier?

Reply:
I don't think it can come any earlier. In order to check this we need to (1) have dropped all empty slices from the list and (2) have gotten the schema from the first entry for comparison with the others, unless we want to implement this using `adjacent_find`, but I find that a lot harder to read.